managed/services/onboarding/tips.go

```diff
@@ -0,0 +1,285 @@
+
+	systemTipIDs map[int64]struct{}
+	userTipIDs   map[int64]struct{}
+
+	onboardingpb.UnimplementedTipServiceServer
+}
+
+var _ onboardingpb.TipServiceServer = (*TipsService)(nil)
+
+func NewTipService(db *reform.DB, inventoryService inventoryService, client grafanaClient) *TipsService {
+	return &TipsService{
+		db:               db,
+		inventoryService: inventoryService,
+		c:                client,
+		systemTipIDs: map[int64]struct{}{
+			InstallPMMServerTipID:    {},
+			InstallPMMClientTipID:    {},
+			ConnectServiceToPMMTipID: {},
+		},
+		userTipIDs: map[int64]struct{}{
+			SeeYourNewAdvisorsTipID: {},
+		},
+	}
+}
+
+func (t *TipsService) GetOnboardingStatus(ctx context.Context, _ *onboardingpb.GetOnboardingStatusRequest) (*onboardingpb.GetOnboardingStatusResponse, error) {
+	userID, err := t.c.GetUserID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	systemTips, err := t.retrieveSystemTips()
+	if err != nil {
+		return nil, err
+	}
+
+	userTips, err := t.retrieveUserTips(userID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &onboardingpb.GetOnboardingStatusResponse{
+		SystemTips: systemTips,
+		UserTips:   userTips,
+	}, nil
+}
+
+func (t *TipsService) retrieveSystemTips() ([]*onboardingpb.TipModel, error) {
+	structs, err := t.db.Querier.SelectAllFrom(models.OnboardingSystemTipTable, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve system tip by id")
+	}
+	tips := make([]*models.OnboardingSystemTip, len(structs))
+	for i, s := range structs {
+		tips[i] = s.(*models.OnboardingSystemTip) //nolint:forcetypeassert
+	}
+
+	for _, tip := range tips {
+		if !tip.IsCompleted {
+			isCompleted, err := t.isSystemTipCompleted(tip.ID)
+			if err != nil {
+				return nil, err
+			}
+			tip.IsCompleted = isCompleted
+
+			err = t.db.Save(tip)
+			if err != nil {
+				return nil, errors.Wrap(err, "cannot save tip info")
+			}
+		}
+	}
+
+	res := make([]*onboardingpb.TipModel, len(tips))
+	for i, tip := range tips {
+		res[i] = &onboardingpb.TipModel{
+			TipId:       tip.ID,
+			IsCompleted: tip.IsCompleted,
+		}
+	}
+	return res, nil
+}
+
+func (t *TipsService) isSystemTipCompleted(tipID int64) (bool, error) {
+	switch tipID {
+	case InstallPMMServerTipID:
+		return true, nil
+	case InstallPMMClientTipID:
+		isCompleted, err := t.isAnyExternalClientConnected()
+		if err != nil {
+			return false, errors.Wrap(err, "Cannot retrieve list of agents to check the status of tip")
+		}
+		return isCompleted, nil
+	case ConnectServiceToPMMTipID:
+		isCompleted, err := t.isAnyServiceConnected()
+		if err != nil {
+			return false, errors.Wrap(err, "Cannot retrieve list of services to check the status of tip")
+		}
+		return isCompleted, nil
+	default:
+		return false, errors.Errorf("system tip doesn't exist: %d", tipID)
+	}
+}
+
+func (t *TipsService) isAnyExternalClientConnected() (bool, error) {
+	pmmServerAgentsByAgentID, err := models.FindAgents(t.db.Querier, models.AgentFilters{
+		PMMAgentID: "pmm-server",
```

Comment:
FYI: we have a constant for this - models.PMMServerAgentID

Reply:
updated